topic_mentions: Fix restriction rule for @-topic mentions.

prakhar1144 opened the pull request:
**First commit:** 

Prep commit to address this comment https://github.com/zulip/zulip/pull/27813#issuecomment-1821332747

<details><summary>@-stream</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/2dcbf4e9-8590-4b23-a7d9-12b9e20187ac"></img>
</details> 

<details><summary>@-everyone</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/222bcbb1-618b-4373-8b13-933f5d6ec3f0"></img>
</details> 

<details><summary>@-all</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/069d12b5-f067-4b6d-927d-46724ef644df"></img>
</details> 

**Second commit:**

Now, the topic wildcard mention follows the following rules:
* If the topic has less than 15 participants , anyone can use @ topic mentions.
* For more than 15, the org setting 'wildcard_mention_policy' determines who can use @ topic mentions.

Fixes part of #27700.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

---

**Screenshots and screen captures:**

**Message Edit:**

<details><summary>Stream wildcard mention</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/8e2320d9-95a2-4b33-9b09-69e9c908913c">
</img>
</details>

<details><summary>Topic wildcard mention</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/2dce81ae-0706-467d-8dbb-424492adcb18">
</img>
</details> 

**Message Send:**

<details><summary>Stream wildcard mention</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c66e0a66-5ae4-42c7-a49d-3db39330b7af">
</img>
</details> 

<details><summary>Topic wildcard mention</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/6f688108-ae1a-4635-be06-31d975260ff1">
</img>
</details> 

**Setting**

<details><summary>Org setting</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/307bd98b-fbe6-46ad-915a-f7d003360b6c">
</img>
</details> 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Posted a [question on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/topic.20wildcard.20mention.20permission/near/1686298)

alya commented:
I think we don't need the "Error editing message:" text in the banner.

alya commented:
Probably clearer to say:

> You do not have permission to use **@topic** mentions in this topic.

That way, a users doesn't have to understand what "wildcard mentions" are.

alya commented:
Probably an improvement for the other mention banner as well:

> You do not have permission to use **@stream** mentions in this stream.
> You do not have permission to use **@all** mentions in this stream.
etc.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1663,14 +1663,7 @@ def get_recent_private_conversations(user_profile: UserProfile) -> Dict[int, Dic
-    if num_subscribers_for_stream_id(stream.id) <= Realm.WILDCARD_MENTION_THRESHOLD:
-        return True
-
+def wildcard_mention_allowed(sender: UserProfile, realm: Realm) -> bool:
```
Let's rename this to something like `large_stream_wildcard_mention_policy_authorizes_user`, maybe with a docstring noting this is a helper for the `topic_wildcard_mention_allowed` and `stream_wildcard_mention_allowed functions -- I want to make clear it's obvious from the codebase that this shouldn't be used directly.

prakhar1144 replied:
Yeah, Renamed it to `wildcard_mention_policy_authorizes_user`. Avoiding the `large_stream` part as it doesn't make sense with topic wildcard in effect. 

`wildcard_mention_policy` is the setting name so the function name clearly conveys that whether that setting permits the user.

Added doctring.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1274,11 +1276,22 @@ def check_update_message(
+                message.sender, topic_participant_count, message.realm
+            ):
+                raise JsonableError(
+                    _("You do not have permission to use topic wildcard mentions in this topic.")
```
We should avoid calling `participants_for_topic`, which is an expensive function, an extra time from this code path; we call `get_recipient_info` in `do_update_message`, which is where the message-send code path gets this ... that call is just after the point of this check. So I think the fix for this issue would be to refactor the `check_update_message/do_update_message` system to put the `get_recipient_info` call in `check_update_message`, and pass it through from there, so that we can do this check using the data that fetches.

It's likely not very important -- editing messages that edit `@topic` mentions will not be super common -- but it is something worth planing to fix. This might be a refactor we should defer to 9.0 (or to a lower-priority follow-up commit late in 8.0). Maybe open an issue for this that we can set as a low-priority release goal?

prakhar1144 replied:
Hm, It'd be 1-2 hours of work. I'll open an issue and backlink here.

timabbott replied:
Yeah that sounds like the right plan for this.

timabbott commented:
Posted a few comments on the engineering; otherwise the backend logic at least looks great.

prakhar1144 commented:
Addressed the review comments and updated the PR description with screenshots.

alya commented:
I'd like the mention in the banner (`@everyone`, etc.) to be bolded.

alya commented:
`@topic` looks to be missing from mention auto-complete in large streams. I thought we'd fixed that? It works on CZO...

alya commented:
Works for me otherwise! I haven't tested the situation where the topic has too many users to notify, as I'm not sure how to conveniently set that up in `dev`.

prakhar1144 commented:
> `@topic` looks to be missing from mention auto-complete in large streams. I thought we'd fixed that? Maybe I'm misremembering.

Works for me

<details><summary>@-topic</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/d7347427-bf63-4057-a4e1-fd408feb27f3">
</img>
</details>

alya commented:
Just to check, did you have wildcard mentions in large streams prohibited for the user you were testing with?

prakhar1144 commented:
> Just to check, did you have wildcard mentions in large streams prohibited for the user you were testing with?

Hm, so the only case where the @-topic doesn't appear in typeahead is:
* Restriction applied but the topic has less than 15 participants. It should appear in that case.

But again the problem is to calculate the participant count in client as the count would vary when messages are still loading for the topic. For one moment, @-topic would be available in the typeahead and when more messages loaded, it'd disappear.

prakhar1144 commented:
> I'd like the mention in the banner (`@everyone`, etc.) to be bolded.

Done. Screenshots updated.

> Works for me otherwise! I haven't tested the situation where the topic has too many users to notify, as I'm not sure how to conveniently set that up in `dev`.

```python manage.py populate_db --extra-users=60``` We get a lot of topics with more than 15 participants.

> https://github.com/zulip/zulip/pull/27813#issuecomment-1822144976

This would be a follow issue to work on

alya commented:
The screenshots look good to me! @timabbott let me know if you need me to do more testing here.

alya commented:
Filed a release goal issue for the follow-up: #27852

timabbott commented on web/src/compose_banner.ts:
```diff
@@ -131,7 +131,14 @@ export function show_error_message(
     classname: string,
     $container: JQuery,
     $bad_input?: JQuery,
+    html_content: boolean = false,
```
This option intentionally does not exist; you're supposed to use partials like `web/templates/compose_banner/automatic_new_visibility_policy_banner.hbs`

I pushed a prep commit adding a comment explaining this.

timabbott commented on web/src/message_edit.js:
```diff
@@ -1070,17 +1070,29 @@ export function save_message_row_edit($row) {
                         compose_banner.CLASSNAMES.generic_compose_error,
                         $container,
+                        undefined,
+                        html_content,
```
I think logic would be just `append_compose_banner_to_banner_list` in the `TOPIC_WILDCARD_MENTION_NOT_ALLOWED` code path, and `compose_banner.show_error_message` in the else clause.

prakhar1144 replied:
Done.

timabbott commented on web/src/transmit.js:
```diff
@@ -70,8 +71,18 @@ export function send_message(request, on_success, error) {
+                } else {
+                    response = channel.xhr_error_message("Error sending message", xhr);
+                }
+                error(response, html_content);
```
This may be a bit more awkward to plumb to the `compose_banner` API. I think one option is to pass to `error` the `code` parameter, and call a new wrapper for `compose_banner.show_error_message` called `compose_banner.show_server_send_failure` which has an extra `server_error_code` parameter, and renders the topic mentions banner if that's appropriate? Not sure, maybe there's a simpler way to do it, but this should work, and I think would not be a bad API change to have a dedicated function for processing server errors.

prakhar1144 replied:
Followed the[ same approach](https://github.com/zulip/zulip/pull/27813#discussion_r1402739099) as discussed for message edit.

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -570,3 +571,15 @@ def __init__(self, msg: str, error_type: str) -> None:
+    @staticmethod
+    @override
+    def msg_format() -> str:
+        return _("You do not have permission to use @topic mentions in this topic.")
```
We should probably document this as a (very minor) API change, or at least a new example failure to document the error code, since other clients may want to do something with it too.

It would not be a bad idea to split the backend changes as a preparatory PR, since that's ready to merge aside from this documentation issue.

prakhar1144 replied:
> We should probably document this as a (very minor) API change, or at least a new example failure to document the error code, since other clients may want to do something with it too.

Done

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1710,13 +1712,21 @@ def check_message(
         raise JsonableError(
-            _("You do not have permission to use wildcard mentions in this stream.")
+            _("You do not have permission to use stream wildcard mentions in this stream.")
         )
```
We should probably turn this into a parallel `StreamWildcardMentionNotAllowedError` thing in `exceptions.py` just so that these work parallel to each other -- I can see mobile and other clients wanting to do special handling of this exception in some situations.

prakhar1144 replied:
Done

timabbott commented:
This is looking close. I would like to see the way we use the `compose_banner` APIs reworked, and a bit of work on API documentation as part of the merging the backend change, but I think this should be good after those tweaks.

prakhar1144 commented:
Addressed the review comments.

<details><summary>update-message</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/3f2bb8d2-d04e-4819-a8d9-3e68819e1fed">
</img>
</details> 

<details><summary>send-message</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/119fdd5e-f633-459a-8fe9-85657653c7a0">
</img>
</details> 

<details><summary>Changelog</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/bfcbed25-f415-4807-957d-d6523d7d2e16">
</img>
</details>

timabbott commented:
I made a few edits to the API documentation -- both to be less verbose in the main changelog, and also to be more specific in the specific documentation for the error code.

I also reworked the `transmit.js` API to avoid skipping the error callback. While I think the code did work correctly originally (because it copied in the content of the error callback for the non-echo case, and echo is disabled when topic mentions are present), I think localizing the error message next to the main error message is a lot more readable.

``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index 00752c6406..c2eb725d76 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -22,14 +22,13 @@ format used by the Zulip server that they are interacting with.
 
 **Feature level 229**
 
-* [`PATCH /messages/{message_id}`](/api/update-message),
-  [`POST /messages`](/api/send-message): Endpoints return specific
-  error responses if one doesn't have the permission to use stream wildcard mentions
-  (`"code": "STREAM_WILDCARD_MENTION_NOT_ALLOWED"`) and topic wildcard mentions
-  (`"code": "STREAM_WILDCARD_MENTION_NOT_ALLOWED"`) while editing or sending messages
-  to a stream and topic, respectively.
-  Previously, `"BAD_REQUEST"` code was returned for stream wildcard mention.
-  Explicit topic wildcard mention restriction was not present earlier.
+* [`PATCH /messages/{message_id}`](/api/update-message), [`POST
+  /messages`](/api/send-message): Topic wildcard mentions involving
+  large numbers of participants are now restricted by
+  `wildcard_mention_policy`. The server now uses the
+  `STREAM_WILDCARD_MENTION_NOT_ALLOWED` and
+  `TOPIC_WILDCARD_MENTION_NOT_ALLOWED` error codes when a message is
+  rejected because of `wildcard_mention_policy`.
 
 **Feature level 228**
 
diff --git a/web/src/compose.js b/web/src/compose.js
index 263df67f8d..574c613f0e 100644
--- a/web/src/compose.js
+++ b/web/src/compose.js
@@ -5,6 +5,7 @@ import $ from "jquery";
 import _ from "lodash";
 
 import render_success_message_scheduled_banner from "../templates/compose_banner/success_message_scheduled_banner.hbs";
+import render_wildcard_mention_not_allowed_error from "../templates/compose_banner/wildcard_mention_not_allowed_error.hbs";
 
 import * as channel from "./channel";
 import * as compose_banner from "./compose_banner";
@@ -180,16 +181,26 @@ export function send_message(request = create_message_object()) {
         send_message_success(request, data);
     }
 
-    function error(response) {
-        // If we're not local echo'ing messages, or if this message was not
-        // locally echoed, show error in compose box
+    function error(response, server_error_code) {
+        // Error callback for failed message send attempts.
         if (!locally_echoed) {
-            compose_banner.show_error_message(
-                response,
-                compose_banner.CLASSNAMES.generic_compose_error,
-                $("#compose_banners"),
-                $("textarea#compose-textarea"),
-            );
+            if (server_error_code === "TOPIC_WILDCARD_MENTION_NOT_ALLOWED") {
+                // The topic wildcard mention permission code path has
+                // a special error.
+                const new_row = render_wildcard_mention_not_allowed_error({
+                    banner_type: compose_banner.ERROR,
+                    classname: compose_banner.CLASSNAMES.wildcards_not_allowed,
+                });
+                compose_banner.append_compose_banner_to_banner_list(new_row, $("#compose_banners"));
+            } else {
+                compose_banner.show_error_message(
+                    response,
+                    compose_banner.CLASSNAMES.generic_compose_error,
+                    $("#compose_banners"),
+                    $("textarea#compose-textarea"),
+                );
+            }
+
             // For messages that were not locally echoed, we're
             // responsible for hiding the compose spinner to restore
             // the compose box so one can send a next message.
diff --git a/web/src/echo.js b/web/src/echo.js
index bd337cf288..cb062dc85f 100644
--- a/web/src/echo.js
+++ b/web/src/echo.js
@@ -96,7 +96,7 @@ function resend_message(message, $row, {on_send_message_success, send_message})
         failed_message_success(message_id);
     }
 
-    function on_error(response) {
+    function on_error(response, _server_error_code) {
         message_send_error(message.id, response);
         setTimeout(() => {
             hide_retry_spinner($row);
diff --git a/web/src/transmit.js b/web/src/transmit.js
index 6811d79200..29180cb4df 100644
--- a/web/src/transmit.js
+++ b/web/src/transmit.js
@@ -1,12 +1,7 @@
 import * as Sentry from "@sentry/browser";
-import $ from "jquery";
-
-import render_wildcard_mention_not_allowed_error from "../templates/compose_banner/wildcard_mention_not_allowed_error.hbs";
 
 import * as blueslip from "./blueslip";
 import * as channel from "./channel";
-import * as compose_banner from "./compose_banner";
-import * as compose_ui from "./compose_ui";
 import {page_params} from "./page_params";
 import * as people from "./people";
 import * as reload from "./reload";
@@ -75,21 +70,8 @@ export function send_message(request, on_success, error) {
                     return;
                 }
 
-                if (xhr.responseJSON?.code === "TOPIC_WILDCARD_MENTION_NOT_ALLOWED") {
-                    const new_row = render_wildcard_mention_not_allowed_error({
-                        banner_type: compose_banner.ERROR,
-                        classname: compose_banner.CLASSNAMES.wildcards_not_allowed,
-                    });
-                    compose_banner.append_compose_banner_to_banner_list(
-                        new_row,
-                        $("#compose_banners"),
-                    );
-                    compose_ui.hide_compose_spinner();
-                    return;
-                }
-
                 const response = channel.xhr_error_message("Error sending message", xhr);
-                error(response);
+                error(response, xhr.responseJSON?.code);
             },
         });
     } finally {
@@ -115,7 +97,7 @@ export function reply_message(opts) {
         //       already handles things like reporting times to the server.)
     }
 
-    function error() {
+    function error(_response, _server_error_code) {
         // TODO: In our current use case, which is widgets, to meaningfully
         //       handle errors, we would want the widget to provide some
         //       kind of callback to us so it can do some appropriate UI.
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 5032dc55c4..82430b4413 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -6413,8 +6413,9 @@ paths:
                             "code": "STREAM_WILDCARD_MENTION_NOT_ALLOWED",
                           }
                         description: |
-                          An example JSON error response for when one doesn't have the permission
-                          to use stream wildcard mention while sending messages in that stream.
+                          An example JSON error response for when the message was rejected because
+                          of the organization's `wildcard_mention_policy` and large number of
+                          subscribers to the stream.
 
                           **Changes**: New in Zulip 8.0 (feature level 229). Previously, this
                           error returned the `"BAD_REQUEST"` code.
@@ -6427,10 +6428,14 @@ paths:
                             "code": "TOPIC_WILDCARD_MENTION_NOT_ALLOWED",
                           }
                         description: |
-                          An example JSON error response for when one doesn't have the permission
-                          to use topic wildcard mention while sending messages in that topic.
+                          An example JSON error response for when the message was rejected because
+                          the message contains a topic wildcard mention, but the user doesn't have
+                          permission to use such a mention in this topic due to the
+                          `wildcard_mention_policy` (and large number of participants in this
+                          specific topic).
 
-                          **Changes**: New in Zulip 8.0 (feature level 229).
+                          **Changes**: New in Zulip 8.0 (feature level 229). Previously,
+                          `wildcard_mention_policy` was not enforced for topic mentions.
   /messages/{message_id}/history:
     get:
       operationId: get-message-history
@@ -7704,8 +7709,9 @@ paths:
                             "code": "STREAM_WILDCARD_MENTION_NOT_ALLOWED",
                           }
                         description: |
-                          An example JSON error response for when one doesn't have the permission to
-                          edit a message to include stream wildcard mention in that stream.
+                          An example JSON error response for when the message was rejected because
+                          of the organization's `wildcard_mention_policy` and large number of
+                          subscribers to the stream.
 
                           **Changes**: New in Zulip 8.0 (feature level 229). Previously, this
                           error returned the `"BAD_REQUEST"` code.
@@ -7718,10 +7724,14 @@ paths:
                             "code": "TOPIC_WILDCARD_MENTION_NOT_ALLOWED",
                           }
                         description: |
-                          An example JSON error response for when one doesn't have the permission to
-                          edit a message to include topic wildcard mention in that topic.
+                          An example JSON error response for when the message was rejected because
+                          the message contains a topic wildcard mention, but the user doesn't have
+                          permission to use such a mention in this topic due to the
+                          `wildcard_mention_policy` (and large number of participants in this
+                          specific topic).
 
-                          **Changes**: New in Zulip 8.0 (feature level 229).
+                          **Changes**: New in Zulip 8.0 (feature level 229). Previously,
+                          `wildcard_mention_policy` was not enforced for topic mentions.
     delete:
       operationId: delete-message
       summary: Delete a message
```

timabbott commented:
Oh, I didn't update the `transmit` tests. Feel free to let it merge once you've fixed that.

timabbott commented:
I did the simplest reasonable test adjustment since it seemed quick and it'll be nice to check this off.
